core/store/src/trie/trie_storage.rs

```diff
@@ -13,25 +13,120 @@ use std::cell::RefCell;
 use std::io::ErrorKind;
 
 #[derive(Clone)]
-pub struct TrieCache(Arc<Mutex<LruCache<CryptoHash, Vec<u8>>>>);
+pub struct SyncTrieCache(Arc<Mutex<TrieCache>>);
+
+struct TrieCache {
+    cache_state: CacheState,
+    shard_cache: LruCache<CryptoHash, Vec<u8>>,
+    chunk_cache: HashMap<CryptoHash, Vec<u8>>,
+}
+
+enum CachePosition {
+    None,
+    ShardCache(Vec<u8>),
+    ChunkCache(Vec<u8>),
+}
+
+enum CacheState {
+    CachingShard,
+    CachingChunk,
+}
+
+pub enum RetrievalCost {
+    Free,
+    Full,
+}
 
 impl TrieCache {
+    fn get_cache_position(&mut self, hash: &CryptoHash) -> CachePosition {
+        match self.chunk_cache.get(hash) {
+            Some(value) => CachePosition::ChunkCache(value.clone()),
```

Comment:
do we want to always clone here?

Reply:
It can be improved, but so far the only usage of this function (`retrieve_raw_bytes`) requires cloning.

On the other hand, `retrieve_raw_bytes` itself can be improved by returning a reference, but I would treat this as a separate issue https://github.com/near/nearcore/issues/6270